Variable selection for graphviz visualization via kwarg

larryshamalama opened the pull request:
Fixes #5527. Some elements are hackish and currently need more work.

+ [x] Do not plot all descendants when specifying an intermediate rv unless the variable is observed (plot observed node)
+ [x] Add `var_names` as kwarg
+ [x] Rename previous `var_names` class instance by `_all_var_names`

larryshamalama committed:
removed all_discrete, added check for same type

larryshamalama committed:
Checking comp_dist.dtype, integration of helper function in .dist()

larryshamalama committed:
graphviz filter function works (for now)

canyon289 commented on pymc/model_graph.py:
```diff
@@ -219,13 +238,18 @@ def make_graph(self, formatting: str = "plain"):
+                try:
+                    graph.edge(parent.replace(":", "&"), child.replace(":", "&"))
+                except AttributeError:
+                    print(child, parent)
```
Why print? Can this be a logging statement?

larryshamalama replied:
Oops, this was left over from when I was debugging 😅 It shouldn't be there

larryshamalama committed:
Remove print statements for debugging:

larryshamalama committed:
Fixed pre-commit checks

larryshamalama committed:
Undo changes in mixture (mixing between branches)

larryshamalama commented:
@michaelosthege Can you have a look at these changes on a model or two? Happy to hear comments. It seems to work on my model in [this notebook](https://github.com/larryshamalama/pymc3-playground/blob/master/notebooks/shortcomings/graphviz-filter-out-vars.ipynb) but we can't see the figures on GitHub

michaelosthege commented on pymc/model_graph.py:
```diff
@@ -219,13 +242,15 @@ def make_graph(self, formatting: str = "plain"):
 
 
-def model_to_graphviz(model=None, *, formatting: str = "plain"):
+def model_to_graphviz(model=None, *, selected_vars=None, formatting: str = "plain"):
```
Let's align it with ArviZ: `var_names`

The ArviZ documentation describes it as:
```
var_names: list[str], optional
    List of variables to plot (defaults to None, which results in all variables plotted).
    Prefix the variables by ~ when you want to exclude them from the plot.
```

larryshamalama replied:
I temporarily use `selected_vars` because `var_names` happens to be an instance variable. Would it be an issue or bad practice to use `var_names` as an argument for the function?

michaelosthege replied:
I don't see a problem with using it as the argument.
You could even rename the `self.var_names` to `self._all_var_names` or so. They should have been private to begin with..

larryshamalama replied:
Done!

michaelosthege reviewed (commented):
It's going into the right direction.
Here's a minimal example that unveils two problems:
1. Selecting nodes causes some of them to change formatting
2. Nodes of observed data have an inverted dependency - this could be tricky.. One can get the desired result by selecting the observed RV, but because of the reversed arrow direction it's counterintuitive.

```python
import pymc as pm
from IPython.display import display

with pm.Model() as pmodel2:
    a = pm.Normal("a")
    b = pm.Normal("b")
    pm.Normal("c", a * b)
    intermediate = pm.Deterministic("intermediate", a + b)
    pred = pm.Deterministic("pred", intermediate * 3)

    obs = pm.ConstantData("obs", 1.75)

    pm.Normal("L", mu=1 + 0.5 * pred, observed=obs)

display(pm.model_to_graphviz(pmodel2))
display(pm.model_to_graphviz(pmodel2, selected_vars=["L"]))
display(pm.model_to_graphviz(pmodel2, selected_vars=["obs"]))
```

![grafik](https://user-images.githubusercontent.com/5894642/159361628-69384da8-2798-4f2e-955c-b8b9e0704be2.png)

michaelosthege commented on pymc/model_graph.py:
```diff
@@ -102,11 +102,31 @@ def get_parents(self, var: TensorVariable) -> Set[VarName]:
         input_map = defaultdict(set)  # type: Dict[str, Set[VarName]]
 
-        for var_name in self.var_names:
+        for var_name in self.vars_to_plot(selected_vars):
```
Commenting here because the diff above has all these codecov warnings...

In addition to the minimal example, I tested it with my big model and the result was missing a lot of variables.
The ancestor determination in `vars_to_plot()` appears to miss a lot of nodes.

The old code started from `self.var_names` which appears to be the complete list. Maybe the way how it's determined in `__init__` is relevant?

__EDIT__: I extended the example to demo this

larryshamalama commented:
Thanks for the pointers and example. Will look into it.

Do you know approximately when the code has been written? My guess is that `_get_ancestors` was written sometime during v3 and doesn't work well anymore

michaelosthege commented:
> Thanks for the pointers and example. Will look into it.
> 
> Do you know approximately when the code has been written? My guess is that `_get_ancestors` was written sometime during v3 and doesn't work well anymore

You can check the git blame: https://github.com/pymc-devs/pymc/blame/main/pymc/model_graph.py

But generally the code works. Also for my biggest model all variables are shown

larryshamalama committed:
Ad hoc method using ancestors works (?)

larryshamalama commented:
I believe that the plots now display desired graphs in your example above. Can you verify?

I am using `ancestors` from [`aesara.graph.basic.py`](https://github.com/aesara-devs/aesara/blob/main/aesara/graph/basic.py#L711) and there seems to many tools in that file that can be beneficial for this file. `get_parents` in `ModelGraph` still seems to work but I'm not sure about `_get_ancestors`, which is why I am using the Aesara one.

michaelosthege commented on pymc/model_graph.py:
```diff
@@ -102,11 +102,42 @@ def get_parents(self, var: TensorVariable) -> Set[VarName]:
+
+            for var_name in selected_vars:
+                if var_name not in self.var_names:
+                    raise ValueError(f"{var_name} is not part of the defined model.")
```
```suggestion
                    raise ValueError(f"{var_name} is not in this model.")
```

michaelosthege replied:
Could also be done with sets: `outside_names = sel_names - set(self.var_names)`

michaelosthege commented on pymc/model_graph.py:
```diff
@@ -102,11 +102,42 @@ def get_parents(self, var: TensorVariable) -> Set[VarName]:
+        if selected_vars is None:
+            return self.var_names
+        else:
+            selected_vars = set(selected_vars)
```
because of the return in the line above you can do
```suggestion
        selected_vars = set(selected_vars)
```
Ideally give it a new name, because it has a different type now: `sel_names: Set[str]`

michaelosthege commented on pymc/model_graph.py:
```diff
@@ -102,11 +102,42 @@ def get_parents(self, var: TensorVariable) -> Set[VarName]:
+                for model_var in self.var_list:
+                    if hasattr(model_var.tag, "observations"):
+                        if model_var.tag.observations == self.model[var_name]:
+                            selected_vars = selected_vars.union({model_var.name})
```
```suggestion
                            selected_vars.add(model_var.name)
```

michaelosthege commented on pymc/model_graph.py:
```diff
@@ -102,11 +102,42 @@ def get_parents(self, var: TensorVariable) -> Set[VarName]:
+
+            for var in selected_ancestors:
+                if hasattr(var.tag, "observations"):
+                    selected_ancestors = selected_ancestors.union({var.tag.observations})
```
```suggestion
                    selected_ancestors.add(var.tag.observations)
```

michaelosthege commented on pymc/model_graph.py:
```diff
@@ -102,11 +102,42 @@ def get_parents(self, var: TensorVariable) -> Set[VarName]:
+            selected_ancestors = [var.name for var in selected_ancestors]
+
+            # ordering of self.var_names is important
+            return selected_ancestors
```
Is is sorted anywhere?
Does the return value need to be sorted?

And you can return the listcomprehension without creating another local variable

larryshamalama replied:
Not sure about sorted, would have to revisit why I wrote this comment again

michaelosthege commented on pymc/model_graph.py:
```diff
@@ -102,11 +102,42 @@ def get_parents(self, var: TensorVariable) -> Set[VarName]:
         return self._filter_parents(var, parents)
 
-    def make_compute_graph(self) -> Dict[str, Set[VarName]]:
+    def vars_to_plot(self, selected_vars=None):
```
```suggestion
    def vars_to_plot(self, var_names: Optional[Iterable[str]]=None) -> List[str]:
```
Reading the code I was confused because `selected_vars` are the _names_, right?

michaelosthege reviewed (commented):
> I believe that the plots now display desired graphs in your example above. Can you verify?

Yes, it's looking good for my big model too! Tomorrow morning I can then finally plot my model and print it on an A3 sheet of paper for the meeting on thursday 😃 So thank you for getting it to work 💯

Maybe a bit unexpected that selecting "L" also shows "obs", but it's probably a good idea to do that. It's a feature.

I added some type hint suggestions, and renaming of variables. My suggestions are incomplete, but I think you get the idea.

You can add tests to https://github.com/pymc-devs/pymc/blob/main/pymc/tests/test_model_graph.py. These tests work by comparing the dict representation of the grpah/plates with the expected outcome.

larryshamalama commented:
Thanks so much for the suggestions! I'll address them one by one. Just wanted you to have the figures for your meeting and work on polishing the code more slowly

larryshamalama committed:
renamed var_names to _all_var_names and selected_vars to var_names

larryshamalama committed:
Add Michael\'s suggestions

larryshamalama committed:
minor comment

larryshamalama commented:
Reminder for myself to add a test

larryshamalama committed:
Added test

larryshamalama commented:
The solution currently is very ad-hoc as I do not understand `_get_ancestors` and `get_parents`. However, the code seems to work. Things to consider for next for this PR:

- Remove variables from model with ~
- Add a test to catch error when variable is not in the model

Things not sure if I should look into more
- refactor `get_parents` so it aligns with [`aesara.graph.basic.ancestors`](https://github.com/aesara-devs/aesara/blob/main/aesara/graph/basic.py#L711)

@michaelosthege can I get your input?

larryshamalama committed:
fix pre-commit

larryshamalama committed:
fixed pre-commit again

larryshamalama committed:
test passing

larryshamalama commented:
Quick comment that I did not forget about this PR, just busy with many other things. There's ongoing work with `networkx` integration, but I need to read up more on that.

larryshamalama committed:
Fixing commit history

larryshamalama committed:
Refactoring graphviz functionality

larryshamalama committed:
Fixing pre-commit issues

larryshamalama committed:
get_parent_names returns a Set instead of List

larryshamalama commented:
I'm revisiting this PR and would like to push it to the finish line after a long hiatus 😅 In summary, this PR:

- adds the possibility to specify a subgraph to plot;
- adds a corresponding test for the above;
- refactors `get_parents` and removes (dangerously...) code that is no longer in use.

I'm hoping that all tests pass. Happy to hear any comments about this progress as it would be worthwhile to address issues #5616 and #5766 shortly afterwards. I imagine that I should add type hints to the methods?

michaelosthege commented:
Type hints are always good 👍

But also make sure to explain the new `var_names` kwarg in the docstring of `model_to_graphviz`.

larryshamalama committed:
Adding test for parents

michaelosthege reviewed (commented):
@larryshamalama we'll only need that docstinrg, then this is good to go!

larryshamalama commented:
Sounds good! I'll fix this later today on the train 😅

larryshamalama committed:
Added docstring and type hints

larryshamalama commented:
I only added the docstring to `pm.model_to_graphviz` since it seemed to be the only user-facing function. I also added type hints and, instead of `str`, I used [`VarName`](https://github.com/pymc-devs/pymc/blob/main/pymc/model_graph.py#L29).

On another note, I refactored a lot of the code since, with `v4`, many of it seemed irrelevant. The hope is that nothing breaks as I rely on functions from`aesara.graph.basic`, but I have a bit of uncertainty there. I'm happy to look into this file if anything else breaks in the future.

michaelosthege commented on pymc/model_graph.py:
```diff
@@ -32,85 +33,64 @@
+    ) -> Dict[VarName, Set[VarName]]:
         """Get map of var_name -> set(input var names) for the model"""
-        input_map = defaultdict(set)  # type: Dict[str, Set[VarName]]
+        input_map = defaultdict(set)  # type: Dict[VarName, Set[VarName]]
```
```suggestion
        input_map: Dict[VarName, Set[VarName]] = defaultdict(set)
```

michaelosthege commented on pymc/model_graph.py:
```diff
@@ -168,18 +149,19 @@ def _make_node(self, var_name, graph, *, formatting: str = "plain"):
         return function([], var, mode="FAST_COMPILE")()
 
-    def get_plates(self):
+    def get_plates(self, var_names: Optional[Iterable[VarName]] = None):
```
```suggestion
    def get_plates(
        self, var_names: Optional[Iterable[VarName]] = None
    ) -> Dict[str, Set[VarName]]:
```

michaelosthege commented on pymc/model_graph.py:
```diff
@@ -168,18 +149,19 @@ def _make_node(self, var_name, graph, *, formatting: str = "plain"):
         Returns
         -------
-        dict: str -> set[str]
+        dict: VarName -> set(VarName)
```
```suggestion
        dict
            Maps plate labels to the set of ``VarName``s inside the plate.
```

michaelosthege reviewed (commented):
I just have a few typing changes, otherwise this looks great 👍

I'll apply my suggestions & fix the pre-commit if needed... ⏳

michaelosthege committed:
Add more type hints

michaelosthege commented:
/pre-commit-run

github-actions[bot] committed:
Run pre-commit

michaelosthege commented:
Thanks @larryshamalama !

I'm sure this will prove to be a very useful feature for the still underappreciated `pm.model_to_graphviz`

larryshamalama commented:
Thanks to you for your suggestions too :)

ricardoV94 commented on pymc/model_graph.py:
```diff
@@ -32,85 +33,64 @@
+            if hasattr(var.tag, "observations"):
+                selected_ancestors.add(var.tag.observations)
+
+        # ordering of self._all_var_names is important
```
I know this has been merged, but what's with this comment? I don't think this will be deterministically ordered since the input comes from a set.

larryshamalama replied:
I'm not sure anymore... it's been a while since I wrote this. I guess we can remove it
